alpha/declcfg/filter/config/v1alpha1/filter.go

```diff
@@ -0,0 +1,280 @@
+package v1alpha1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"maps"
+	"strings"
+
+	mmsemver "github.com/Masterminds/semver/v3"
+	blangsemver "github.com/blang/semver/v4"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+	"github.com/operator-framework/operator-registry/alpha/declcfg/filter"
+	"github.com/operator-framework/operator-registry/alpha/model"
+)
+
+type filterOptions struct {
+	Log *logrus.Entry
+}
+
+type FilterOption func(*filterOptions)
+
+type filterer struct {
+	pkgConfigs map[string]Package
+	keeps      map[string]sets.Set[string]
+	opts       filterOptions
+}
+
+func WithLogger(log *logrus.Entry) FilterOption {
+	return func(opts *filterOptions) {
+		opts.Log = log
+	}
+}
+
+func nullLogger() *logrus.Entry {
+	l := logrus.New()
+	l.SetOutput(io.Discard)
+	return logrus.NewEntry(l)
+}
+
+func NewFilterer(config *FilterConfiguration, filterOpts ...FilterOption) filter.Filterer {
+	opts := filterOptions{
+		Log: nullLogger(),
+	}
+	for _, opt := range filterOpts {
+		opt(&opts)
+	}
+	pkgConfigs := make(map[string]Package, len(config.Packages))
+	keeps := make(map[string]sets.Set[string], len(config.Packages))
+	for _, pkg := range config.Packages {
+		pkgConfigs[pkg.Name] = pkg
+		channels := sets.New[string]()
+		for _, ch := range pkg.Channels {
+			channels.Insert(ch.Name)
+		}
+		keeps[pkg.Name] = channels
+	}
+	return &filterer{
+		pkgConfigs: pkgConfigs,
+		keeps:      keeps,
+		opts:       opts,
+	}
+}
+
+func (f *filterer) FilterCatalog(_ context.Context, fbc *declcfg.DeclarativeConfig) (*declcfg.DeclarativeConfig, error) {
```

Comment:
I've seen `context.Context` get added to so many functions at this point that I pretty much add it defensively for functions I expect:
- could be long running
- could need to call other functions that accept a context

Reply:
Ok, I'm going to leave it in. One possibility I thought of is an interaction with an OCI registry to get the labels of bundle images and filter based on them. I'm considering adding a context to the meta filter signature as well for the same reason. WDYT?

Reply:
IMHO, we should have it for consistency reasons (most of the declcfg code follows this convention), but it's also proven really helpful when we refactored for parallelism previously, and the downside ... ?

Reply:
I think it's fine to include if we are anticipating the addition of a filtering method that will use it in the near future. Just calling out that it may be a premature optimization that could be added in the future when needed.

I also am not strongly opinionated here so if folks feel strongly we should keep it - fine by me